File: src/harness/sourceMapRecorder.ts
namespace Harness.SourceMapRecorder {

    interface SourceMapSpanWithDecodeErrors {
        sourceMapSpan: ts.Mapping;
        decodeErrors: string[] | undefined;
    }

    namespace SourceMapDecoder {
        let sourceMapMappings: string;
        let decodingIndex: number;
        let mappings: ts.MappingsDecoder | undefined;

        export interface DecodedMapping {
            sourceMapSpan: ts.Mapping;
            error?: string;
        }

        export function initializeSourceMapDecoding(sourceMap: ts.RawSourceMap) {
            decodingIndex = 0;
            sourceMapMappings = sourceMap.mappings;
            mappings = ts.decodeMappings(sourceMap.mappings);
        }

        export function decodeNextEncodedSourceMapSpan(): DecodedMapping {
            if (!mappings) return ts.Debug.fail("not initialized");
            const result = mappings.next();
            if (result.done) return { error: mappings.error || "No encoded entry found", sourceMapSpan: mappings.state };
            return { sourceMapSpan: result.value };
        }

        export function hasCompletedDecoding() {
            if (!mappings) return ts.Debug.fail("not initialized");
            return mappings.pos === sourceMapMappings.length;
        }

        export function getRemainingDecodeString() {
            return sourceMapMappings.substr(decodingIndex);
        }
    }

    namespace SourceMapSpanWriter {
        let sourceMapRecorder: Compiler.WriterAggregator;
        let sourceMapSources: string[];
        let sourceMapNames: string[] | null | undefined;

        let jsFile: documents.TextDocument;
        let jsLineMap: readonly number[];
        let tsCode: string;
        let tsLineMap: number[];

        let spansOnSingleLine: SourceMapSpanWithDecodeErrors[];
        let prevWrittenSourcePos: number;
        let nextJsLineToWrite: number;
        let spanMarkerContinues: boolean;

        export function initializeSourceMapSpanWriter(sourceMapRecordWriter: Compiler.WriterAggregator, sourceMap: ts.RawSourceMap, currentJsFile: documents.TextDocument) {
            sourceMapRecorder = sourceMapRecordWriter;
            sourceMapSources = sourceMap.sources;
            sourceMapNames = sourceMap.names;

            jsFile = currentJsFile;
            jsLineMap = jsFile.lineStarts;

            spansOnSingleLine = [];
            prevWrittenSourcePos = 0;
            nextJsLineToWrite = 0;
            spanMarkerContinues = false;

            SourceMapDecoder.initializeSourceMapDecoding(sourceMap);
            sourceMapRecorder.WriteLine("===================================================================");
            sourceMapRecorder.WriteLine("JsFile: " + sourceMap.file);
            sourceMapRecorder.WriteLine("mapUrl: " + ts.tryGetSourceMappingURL(ts.getLineInfo(jsFile.text, jsLineMap)));
            sourceMapRecorder.WriteLine("sourceRoot: " + sourceMap.sourceRoot);
            sourceMapRecorder.WriteLine("sources: " + sourceMap.sources);
            if (sourceMap.sourcesContent) {
                sourceMapRecorder.WriteLine("sourcesContent: " + JSON.stringify(sourceMap.sourcesContent));
            }
            sourceMapRecorder.WriteLine("===================================================================");
        }

        function getSourceMapSpanString(mapEntry: ts.Mapping, getAbsentNameIndex?: boolean) {
            let mapString = "Emitted(" + (mapEntry.generatedLine + 1) + ", " + (mapEntry.generatedCharacter + 1) + ")";
            if (ts.isSourceMapping(mapEntry)) {
                mapString += " Source(" + (mapEntry.sourceLine + 1) + ", " + (mapEntry.sourceCharacter + 1) + ") + SourceIndex(" + mapEntry.sourceIndex + ")";
                if (mapEntry.nameIndex! >= 0 && mapEntry.nameIndex! < sourceMapNames!.length) {
                    mapString += " name (" + sourceMapNames![mapEntry.nameIndex!] + ")";
                }
                else {
                    if ((mapEntry.nameIndex && mapEntry.nameIndex !== -1) || getAbsentNameIndex) {
                        mapString += " nameIndex (" + mapEntry.nameIndex + ")";
                    }
                }
            }

            return mapString;
        }

        export function recordSourceMapSpan(sourceMapSpan: ts.Mapping) {
            // verify the decoded span is same as the new span
            const decodeResult = SourceMapDecoder.decodeNextEncodedSourceMapSpan();
            let decodeErrors: string[] | undefined;
            if (typeof decodeResult.error === "string" || !ts.sameMapping(decodeResult.sourceMapSpan, sourceMapSpan)) {
                if (decodeResult.error) {
                    decodeErrors = ["!!^^ !!^^ There was decoding error in the sourcemap at this location: " + decodeResult.error];
                }
                else {
                    decodeErrors = ["!!^^ !!^^ The decoded span from sourcemap's mapping entry does not match what was encoded for this span:"];
                }
                decodeErrors.push("!!^^ !!^^ Decoded span from sourcemap's mappings entry: " + getSourceMapSpanString(decodeResult.sourceMapSpan, /*getAbsentNameIndex*/ true) + " Span encoded by the emitter:" + getSourceMapSpanString(sourceMapSpan, /*getAbsentNameIndex*/ true));
            }

            if (spansOnSingleLine.length && spansOnSingleLine[0].sourceMapSpan.generatedLine !== sourceMapSpan.generatedLine) {
                // On different line from the one that we have been recording till now,
                writeRecordedSpans();
                spansOnSingleLine = [];
            }
            spansOnSingleLine.push({ sourceMapSpan, decodeErrors });
        }

        export function recordNewSourceFileSpan(sourceMapSpan: ts.Mapping, newSourceFileCode: string) {
            let continuesLine = false;
            if (spansOnSingleLine.length > 0 && spansOnSingleLine[0].sourceMapSpan.generatedCharacter === sourceMapSpan.generatedLine) {
                writeRecordedSpans();
                spansOnSingleLine = [];
                nextJsLineToWrite--; // walk back one line to reprint the line
                continuesLine = true;
            }

            recordSourceMapSpan(sourceMapSpan);

            assert.isTrue(spansOnSingleLine.length === 1);
            sourceMapRecorder.WriteLine("-------------------------------------------------------------------");
            sourceMapRecorder.WriteLine("emittedFile:" + jsFile.file + (continuesLine ? ` (${sourceMapSpan.generatedLine + 1}, ${sourceMapSpan.generatedCharacter + 1})` : ""));
            sourceMapRecorder.WriteLine("sourceFile:" + sourceMapSources[spansOnSingleLine[0].sourceMapSpan.sourceIndex!]);
            sourceMapRecorder.WriteLine("-------------------------------------------------------------------");

            tsLineMap = ts.computeLineStarts(newSourceFileCode);
            tsCode = newSourceFileCode;
            prevWrittenSourcePos = 0;
        }

        export function close() {
            // Write the lines pending on the single line
            writeRecordedSpans();

            if (!SourceMapDecoder.hasCompletedDecoding()) {
                sourceMapRecorder.WriteLine("!!!! **** There are more source map entries in the sourceMap's mapping than what was encoded");
                sourceMapRecorder.WriteLine("!!!! **** Remaining decoded string: " + SourceMapDecoder.getRemainingDecodeString());

            }

            // write remaining js lines
            writeJsFileLines(jsLineMap.length);
        }

        function getTextOfLine(line: number, lineMap: readonly number[], code: string) {
            const startPos = lineMap[line];
            const endPos = lineMap[line + 1];
            const text = code.substring(startPos, endPos);
            return line === 0 ? Utils.removeByteOrderMark(text) : text;
        }

        function writeJsFileLines(endJsLine: number) {
            for (; nextJsLineToWrite < endJsLine; nextJsLineToWrite++) {
                sourceMapRecorder.Write(">>>" + getTextOfLine(nextJsLineToWrite, jsLineMap, jsFile.text));
            }
        }

        function writeRecordedSpans() {
            const markerIds: string[] = [];

            function getMarkerId(markerIndex: number) {
                let markerId = "";
                if (spanMarkerContinues) {
                    assert.isTrue(markerIndex === 0);
                    markerId = "1->";
                }
                else {
                    markerId = "" + (markerIndex + 1);
                    if (markerId.length < 2) {
                        markerId = markerId + " ";
                    }
                    markerId += ">";
                }
                return markerId;
            }

            let prevEmittedCol!: number;
            function iterateSpans(fn: (currentSpan: SourceMapSpanWithDecodeErrors, index: number) => void) {
                prevEmittedCol = 0;
                for (let i = 0; i < spansOnSingleLine.length; i++) {
                    fn(spansOnSingleLine[i], i);
                    prevEmittedCol = spansOnSingleLine[i].sourceMapSpan.generatedCharacter;
                }
            }

            function writeSourceMapIndent(indentLength: number, indentPrefix: string) {
                sourceMapRecorder.Write(indentPrefix);
                for (let i = 0; i < indentLength; i++) {
                    sourceMapRecorder.Write(" ");
                }
            }

            function writeSourceMapMarker(currentSpan: SourceMapSpanWithDecodeErrors, index: number, endColumn = currentSpan.sourceMapSpan.generatedCharacter, endContinues = false) {
                const markerId = getMarkerId(index);
                markerIds.push(markerId);

                writeSourceMapIndent(prevEmittedCol, markerId);

                for (let i = prevEmittedCol; i < endColumn; i++) {
                    sourceMapRecorder.Write("^");
                }
                if (endContinues) {
                    sourceMapRecorder.Write("->");
                }
                sourceMapRecorder.WriteLine("");
                spanMarkerContinues = endContinues;
            }

            function writeSourceMapSourceText(currentSpan: SourceMapSpanWithDecodeErrors, index: number) {
                const sourcePos = tsLineMap[currentSpan.sourceMapSpan.sourceLine!] + (currentSpan.sourceMapSpan.sourceCharacter!);
                let sourceText = "";
                if (prevWrittenSourcePos < sourcePos) {
                    // Position that goes forward, get text
                    sourceText = tsCode.substring(prevWrittenSourcePos, sourcePos);
                }

                if (currentSpan.decodeErrors) {
                    // If there are decode errors, write
                    for (const decodeError of currentSpan.decodeErrors) {
                        writeSourceMapIndent(prevEmittedCol, markerIds[index]);
                        sourceMapRecorder.WriteLine(decodeError);
                    }
                }

                const tsCodeLineMap = ts.computeLineStarts(sourceText);
                for (let i = 0; i < tsCodeLineMap.length; i++) {
                    writeSourceMapIndent(prevEmittedCol, i === 0 ? markerIds[index] : "  >");
                    sourceMapRecorder.Write(getTextOfLine(i, tsCodeLineMap, sourceText));
                    if (i === tsCodeLineMap.length - 1) {
                        sourceMapRecorder.WriteLine("");
                    }
                }

                prevWrittenSourcePos = sourcePos;
            }

            function writeSpanDetails(currentSpan: SourceMapSpanWithDecodeErrors, index: number) {
                sourceMapRecorder.WriteLine(markerIds[index] + getSourceMapSpanString(currentSpan.sourceMapSpan));
            }

            if (spansOnSingleLine.length) {
                const currentJsLine = spansOnSingleLine[0].sourceMapSpan.generatedLine;

                // Write js line
                writeJsFileLines(currentJsLine + 1);

                // Emit markers
                iterateSpans(writeSourceMapMarker);

                const jsFileText = getTextOfLine(currentJsLine + 1, jsLineMap, jsFile.text);
                if (prevEmittedCol < jsFileText.length - 1) {
                    // There is remaining text on this line that will be part of next source span so write marker that continues
                    writeSourceMapMarker(/*currentSpan*/ undefined!, spansOnSingleLine.length, /*endColumn*/ jsFileText.length - 1, /*endContinues*/ true); // TODO: GH#18217
                }

                // Emit Source text
                iterateSpans(writeSourceMapSourceText);

                // Emit column number etc
                iterateSpans(writeSpanDetails);

                sourceMapRecorder.WriteLine("---");
            }
        }
    }

    export function getSourceMapRecord(sourceMapDataList: readonly ts.SourceMapEmitResult[], program: ts.Program, jsFiles: readonly documents.TextDocument[], declarationFiles: readonly documents.TextDocument[]) {
        const sourceMapRecorder = new Compiler.WriterAggregator();

        for (let i = 0; i < sourceMapDataList.length; i++) {
            const sourceMapData = sourceMapDataList[i];
            let prevSourceFile: ts.SourceFile | undefined;
            let currentFile: documents.TextDocument;
            if (ts.endsWith(sourceMapData.sourceMap.file, ts.Extension.Dts)) {
                if (sourceMapDataList.length > jsFiles.length) {
                    currentFile = declarationFiles[Math.floor(i / 2)]; // When both kinds of source map are present, they alternate js/dts
                }
                else {
                    currentFile = declarationFiles[i];
                }
            }
            else {
                if (sourceMapDataList.length > jsFiles.length) {
                    currentFile = jsFiles[Math.floor(i / 2)];
                }
                else {
                    currentFile = jsFiles[i];
                }
            }

            SourceMapSpanWriter.initializeSourceMapSpanWriter(sourceMapRecorder, sourceMapData.sourceMap, currentFile);
            const mapper = ts.decodeMappings(sourceMapData.sourceMap.mappings);
            for (let iterResult = mapper.next(); !iterResult.done; iterResult = mapper.next()) {
                const decodedSourceMapping = iterResult.value;
                const currentSourceFile = ts.isSourceMapping(decodedSourceMapping)
                    ? program.getSourceFile(sourceMapData.inputSourceFileNames[decodedSourceMapping.sourceIndex])
                    : undefined;
                if (currentSourceFile !== prevSourceFile) {
                    if (currentSourceFile) {
                        SourceMapSpanWriter.recordNewSourceFileSpan(decodedSourceMapping, currentSourceFile.text);
                    }
                    prevSourceFile = currentSourceFile;
                }
                else {
                    SourceMapSpanWriter.recordSourceMapSpan(decodedSourceMapping);
                }
            }
            SourceMapSpanWriter.close(); // If the last spans werent emitted, emit them
        }
        sourceMapRecorder.Close();
        return sourceMapRecorder.lines.join("\r\n");
    }

    export function getSourceMapRecordWithSystem(sys: ts.System, sourceMapFile: string) {
        const sourceMapRecorder = new Compiler.WriterAggregator();
        let prevSourceFile: documents.TextDocument | undefined;
        const files = new ts.Map<string, documents.TextDocument>();
        const sourceMap = ts.tryParseRawSourceMap(sys.readFile(sourceMapFile, "utf8")!);
        if (sourceMap) {
            const mapDirectory = ts.getDirectoryPath(sourceMapFile);
            const sourceRoot = sourceMap.sourceRoot ? ts.getNormalizedAbsolutePath(sourceMap.sourceRoot, mapDirectory) : mapDirectory;
            const generatedAbsoluteFilePath = ts.getNormalizedAbsolutePath(sourceMap.file, mapDirectory);
            const sourceFileAbsolutePaths = sourceMap.sources.map(source => ts.getNormalizedAbsolutePath(source, sourceRoot));
            const currentFile = getFile(generatedAbsoluteFilePath);

            SourceMapSpanWriter.initializeSourceMapSpanWriter(sourceMapRecorder, sourceMap, currentFile);
            const mapper = ts.decodeMappings(sourceMap.mappings);
            for (let iterResult = mapper.next(); !iterResult.done; iterResult = mapper.next()) {
                const decodedSourceMapping = iterResult.value;
                const currentSourceFile = ts.isSourceMapping(decodedSourceMapping)
                    ? getFile(sourceFileAbsolutePaths[decodedSourceMapping.sourceIndex])
                    : undefined;
                if (currentSourceFile !== prevSourceFile) {
                    if (currentSourceFile) {
                        SourceMapSpanWriter.recordNewSourceFileSpan(decodedSourceMapping, currentSourceFile.text);
                    }
                    prevSourceFile = currentSourceFile;
                }
                else {
                    SourceMapSpanWriter.recordSourceMapSpan(decodedSourceMapping);
                }
            }
            SourceMapSpanWriter.close(); // If the last spans werent emitted, emit them
        }
        sourceMapRecorder.Close();
        return sourceMapRecorder.lines.join("\r\n");

        function getFile(path: string) {
            const existing = files.get(path);
            if (existing) return existing;
            const value = new documents.TextDocument(path, sys.readFile(path, "utf8")!);
            files.set(path, value);
            return value;
        }
    }
}

File: tsc/internal/testutil/harnessutil/sourcemap_recorder.go
package harnessutil

import (
	"encoding/json"
	"errors"
	"fmt"
	"strconv"
	"strings"

	"github.com/microsoft/typescript-go/internal/core"
	"github.com/microsoft/typescript-go/internal/sourcemap"
	"github.com/microsoft/typescript-go/internal/stringutil"
)

type writerAggregator struct {
	strings.Builder
}

func (w *writerAggregator) WriteStringF(format string, args ...any) {
	w.WriteString(fmt.Sprintf(format, args...))
}

func (w *writerAggregator) WriteLine(s string) {
	w.WriteString(s + "\r\n")
}

func (w *writerAggregator) WriteLineF(format string, args ...any) {
	w.WriteStringF(format+"\r\n", args...)
}

type sourceMapSpanWithDecodeErrors struct {
	sourceMapSpan *sourcemap.Mapping
	decodeErrors  []string
}

type decodedMapping struct {
	sourceMapSpan *sourcemap.Mapping
	error         error
}

type sourceMapDecoder struct {
	sourceMapMappings string
	mappings          *sourcemap.MappingsDecoder
}

func newSourceMapDecoder(sourceMap *sourcemap.RawSourceMap) *sourceMapDecoder {
	return &sourceMapDecoder{
		sourceMapMappings: sourceMap.Mappings,
		mappings:          sourcemap.DecodeMappings(sourceMap.Mappings),
	}
}

func (d *sourceMapDecoder) decodeNextEncodedSourceMapSpan() *decodedMapping {
	value, done := d.mappings.Next()
	if done {
		mapping := &decodedMapping{
			error:         d.mappings.Error(),
			sourceMapSpan: d.mappings.State(),
		}
		if mapping.error == nil {
			mapping.error = errors.New("No encoded entry found")
		}
		return mapping
	}
	return &decodedMapping{sourceMapSpan: value}
}

func (d *sourceMapDecoder) hasCompletedDecoding() bool {
	return d.mappings.Pos() == len(d.sourceMapMappings)
}

func (d *sourceMapDecoder) getRemainingDecodeString() string {
	return d.sourceMapMappings[d.mappings.Pos():]
}

type sourceMapSpanWriter struct {
	sourceMapRecorder    *writerAggregator
	sourceMapSources     []string
	sourceMapNames       []string
	jsFile               *TestFile
	jsLineMap            []core.TextPos
	tsCode               string
	tsLineMap            []core.TextPos
	spansOnSingleLine    []sourceMapSpanWithDecodeErrors
	prevWrittenSourcePos int
	nextJsLineToWrite    int
	spanMarkerContinues  bool
	sourceMapDecoder     *sourceMapDecoder
}

func newSourceMapSpanWriter(sourceMapRecorder *writerAggregator, sourceMap *sourcemap.RawSourceMap, jsFile *TestFile) *sourceMapSpanWriter {
	writer := &sourceMapSpanWriter{
		sourceMapRecorder:    sourceMapRecorder,
		sourceMapSources:     sourceMap.Sources,
		sourceMapNames:       sourceMap.Names,
		jsFile:               jsFile,
		jsLineMap:            core.ComputeLineStarts(jsFile.Content),
		spansOnSingleLine:    make([]sourceMapSpanWithDecodeErrors, 0),
		prevWrittenSourcePos: 0,
		nextJsLineToWrite:    0,
		spanMarkerContinues:  false,
		sourceMapDecoder:     newSourceMapDecoder(sourceMap),
	}

	sourceMapRecorder.WriteLine("===================================================================")
	sourceMapRecorder.WriteLineF("JsFile: %s", sourceMap.File)
	sourceMapRecorder.WriteLineF("mapUrl: %s", sourcemap.TryGetSourceMappingURL(sourcemap.GetLineInfo(jsFile.Content, writer.jsLineMap)))
	sourceMapRecorder.WriteLineF("sourceRoot: %s", sourceMap.SourceRoot)
	sourceMapRecorder.WriteLineF("sources: %s", strings.Join(sourceMap.Sources, ","))
	if len(sourceMap.SourcesContent) > 0 {
		content, err := json.Marshal(sourceMap.SourcesContent)
		if err != nil {
			panic(err)
		}
		sourceMapRecorder.WriteLineF("sourcesContent: %s", content)
	}
	sourceMapRecorder.WriteLine("===================================================================")
	return writer
}

func (w *sourceMapSpanWriter) getSourceMapSpanString(mapEntry *sourcemap.Mapping, getAbsentNameIndex bool) string {
	var mapString writerAggregator
	mapString.WriteStringF("Emitted(%d, %d)", mapEntry.GeneratedLine+1, mapEntry.GeneratedCharacter+1)
	if mapEntry.IsSourceMapping() {
		mapString.WriteStringF(" Source(%d, %d) + SourceIndex(%d)", mapEntry.SourceLine+1, mapEntry.SourceCharacter+1, mapEntry.SourceIndex)
		if mapEntry.NameIndex >= 0 && int(mapEntry.NameIndex) < len(w.sourceMapNames) {
			mapString.WriteStringF(" name (%s)", w.sourceMapNames[mapEntry.NameIndex])
		} else {
			if mapEntry.NameIndex != sourcemap.MissingName || getAbsentNameIndex {
				mapString.WriteStringF(" nameIndex (%d)", mapEntry.NameIndex)
			}
		}
	}
	return mapString.String()
}

func (w *sourceMapSpanWriter) recordSourceMapSpan(sourceMapSpan *sourcemap.Mapping) {
	// verify the decoded span is same as the new span
	decodeResult := w.sourceMapDecoder.decodeNextEncodedSourceMapSpan()
	var decodeErrors []string
	if decodeResult.error != nil || !decodeResult.sourceMapSpan.Equals(sourceMapSpan) {
		if decodeResult.error != nil {
			decodeErrors = []string{"!!^^ !!^^ There was decoding error in the sourcemap at this location: " + decodeResult.error.Error()}
		} else {
			decodeErrors = []string{"!!^^ !!^^ The decoded span from sourcemap's mapping entry does not match what was encoded for this span:"}
		}
		decodeErrors = append(decodeErrors,
			"!!^^ !!^^ Decoded span from sourcemap's mappings entry: "+
				w.getSourceMapSpanString(decodeResult.sourceMapSpan, true /*getAbsentNameIndex*/)+
				" Span encoded by the emitter:"+
				w.getSourceMapSpanString(sourceMapSpan, true /*getAbsentNameIndex*/),
		)
	}

	if len(w.spansOnSingleLine) > 0 && w.spansOnSingleLine[0].sourceMapSpan.GeneratedLine != sourceMapSpan.GeneratedLine {
		// On different line from the one that we have been recording till now,
		w.writeRecordedSpans()
		w.spansOnSingleLine = nil
	}
	w.spansOnSingleLine = append(w.spansOnSingleLine, sourceMapSpanWithDecodeErrors{
		sourceMapSpan: sourceMapSpan,
		decodeErrors:  decodeErrors,
	})
}

func (w *sourceMapSpanWriter) recordNewSourceFileSpan(sourceMapSpan *sourcemap.Mapping, newSourceFileCode string) {
	continuesLine := false
	if len(w.spansOnSingleLine) > 0 && w.spansOnSingleLine[0].sourceMapSpan.GeneratedCharacter == sourceMapSpan.GeneratedLine { // !!! char == line seems like a bug in Strada?
		w.writeRecordedSpans()
		w.spansOnSingleLine = nil
		w.nextJsLineToWrite-- // walk back one line to reprint the line
		continuesLine = true
	}

	w.recordSourceMapSpan(sourceMapSpan)

	if len(w.spansOnSingleLine) != 1 {
		panic("expected a single span")
	}

	w.sourceMapRecorder.WriteLine("-------------------------------------------------------------------")
	if continuesLine {
		w.sourceMapRecorder.WriteLineF("emittedFile:%s (%d, %d)", w.jsFile.UnitName, sourceMapSpan.GeneratedLine+1, sourceMapSpan.GeneratedCharacter+1)
	} else {
		w.sourceMapRecorder.WriteLineF("emittedFile:%s", w.jsFile.UnitName)
	}
	w.sourceMapRecorder.WriteLineF("sourceFile:%s", w.sourceMapSources[w.spansOnSingleLine[0].sourceMapSpan.SourceIndex])
	w.sourceMapRecorder.WriteLine("-------------------------------------------------------------------")

	w.tsLineMap = core.ComputeLineStarts(newSourceFileCode)
	w.tsCode = newSourceFileCode
	w.prevWrittenSourcePos = 0
}

func (w *sourceMapSpanWriter) close() {
	// Write the lines pending on the single line
	w.writeRecordedSpans()

	if !w.sourceMapDecoder.hasCompletedDecoding() {
		w.sourceMapRecorder.WriteLine("!!!! **** There are more source map entries in the sourceMap's mapping than what was encoded")
		w.sourceMapRecorder.WriteLineF("!!!! **** Remaining decoded string: %s", w.sourceMapDecoder.getRemainingDecodeString())
	}

	// write remaining js lines
	w.writeJsFileLines(len(w.jsLineMap))
}

func (w *sourceMapSpanWriter) getTextOfLine(line int, lineMap []core.TextPos, code string) string {
	startPos := lineMap[line]
	var endPos core.TextPos
	if line+1 < len(lineMap) {
		endPos = lineMap[line+1]
	} else {
		endPos = core.TextPos(len(code))
	}
	text := code[startPos:endPos]
	if line == 0 {
		return stringutil.RemoveByteOrderMark(text)
	}
	// return line == 0 ? Utils.removeByteOrderMark(text) : text;
	return text
}

func (w *sourceMapSpanWriter) writeJsFileLines(endJsLine int) {
	for ; w.nextJsLineToWrite < endJsLine; w.nextJsLineToWrite++ {
		w.sourceMapRecorder.WriteStringF(">>>%s", w.getTextOfLine(w.nextJsLineToWrite, w.jsLineMap, w.jsFile.Content))
	}
}

func (w *sourceMapSpanWriter) writeRecordedSpans() {
	recordedSpanWriter := recordedSpanWriter{w: w}
	recordedSpanWriter.writeRecordedSpans()
}

type recordedSpanWriter struct {
	markerIds      []string
	prevEmittedCol int
	w              *sourceMapSpanWriter
}

func (sw *recordedSpanWriter) getMarkerId(markerIndex int) string {
	markerId := ""
	if sw.w.spanMarkerContinues {
		if markerIndex != 0 {
			panic("expected markerIndex to be 0")
		}
		markerId = "1->"
	} else {
		markerId = strconv.Itoa(markerIndex + 1)
		if len(markerId) < 2 {
			markerId += " "
		}
		markerId += ">"
	}
	return markerId
}

func (sw *recordedSpanWriter) iterateSpans(fn func(currentSpan *sourceMapSpanWithDecodeErrors, index int)) {
	sw.prevEmittedCol = 0
	for i := range len(sw.w.spansOnSingleLine) {
		fn(&sw.w.spansOnSingleLine[i], i)
		sw.prevEmittedCol = sw.w.spansOnSingleLine[i].sourceMapSpan.GeneratedCharacter
	}
}

func (sw *recordedSpanWriter) writeSourceMapIndent(indentLength int, indentPrefix string) {
	sw.w.sourceMapRecorder.WriteString(indentPrefix)
	for range indentLength {
		sw.w.sourceMapRecorder.WriteString(" ")
	}
}

func (sw *recordedSpanWriter) writeSourceMapMarker(currentSpan *sourceMapSpanWithDecodeErrors, index int) {
	sw.writeSourceMapMarkerEx(currentSpan, index, currentSpan.sourceMapSpan.GeneratedCharacter, false /*endContinues*/)
}

func (sw *recordedSpanWriter) writeSourceMapMarkerEx(currentSpan *sourceMapSpanWithDecodeErrors, index int, endColumn int, endContinues bool) {
	markerId := sw.getMarkerId(index)
	sw.markerIds = append(sw.markerIds, markerId)
	sw.writeSourceMapIndent(sw.prevEmittedCol, markerId)
	for i := sw.prevEmittedCol; i < endColumn; i++ {
		sw.w.sourceMapRecorder.WriteString("^")
	}
	if endContinues {
		sw.w.sourceMapRecorder.WriteString("->")
	}
	sw.w.sourceMapRecorder.WriteLine("")
	sw.w.spanMarkerContinues = endContinues
}

func (sw *recordedSpanWriter) writeSourceMapSourceText(currentSpan *sourceMapSpanWithDecodeErrors, index int) {
	sourcePos := int(sw.w.tsLineMap[currentSpan.sourceMapSpan.SourceLine]) + currentSpan.sourceMapSpan.SourceCharacter
	var sourceText string
	if sw.w.prevWrittenSourcePos < sourcePos {
		// Position that goes forward, get text
		sourceText = sw.w.tsCode[sw.w.prevWrittenSourcePos:sourcePos]
	}

	// If there are decode errors, write
	for _, decodeError := range currentSpan.decodeErrors {
		sw.writeSourceMapIndent(sw.prevEmittedCol, sw.markerIds[index])
		sw.w.sourceMapRecorder.WriteLine(decodeError)
	}

	tsCodeLineMap := core.ComputeLineStarts(sourceText)
	for i := range tsCodeLineMap {
		if i == 0 {
			sw.writeSourceMapIndent(sw.prevEmittedCol, sw.markerIds[index])
		} else {
			sw.writeSourceMapIndent(sw.prevEmittedCol, "  >")
		}
		sw.w.sourceMapRecorder.WriteString(sw.w.getTextOfLine(i, tsCodeLineMap, sourceText))
		if i == len(tsCodeLineMap)-1 {
			sw.w.sourceMapRecorder.WriteLine("")
		}
	}

	sw.w.prevWrittenSourcePos = sourcePos
}

func (sw *recordedSpanWriter) writeSpanDetails(currentSpan *sourceMapSpanWithDecodeErrors, index int) {
	sw.w.sourceMapRecorder.WriteLineF("%s%s", sw.markerIds[index], sw.w.getSourceMapSpanString(currentSpan.sourceMapSpan, false /*getAbsentNameIndex*/))
}

func (sw *recordedSpanWriter) writeRecordedSpans() {
	w := sw.w
	writeSourceMapMarker := sw.writeSourceMapMarker
	writeSourceMapSourceText := sw.writeSourceMapSourceText
	writeSpanDetails := sw.writeSpanDetails

	if len(w.spansOnSingleLine) > 0 {
		currentJsLine := w.spansOnSingleLine[0].sourceMapSpan.GeneratedLine

		// Write js line
		w.writeJsFileLines(currentJsLine + 1)

		// Emit markers
		sw.iterateSpans(writeSourceMapMarker)

		jsFileText := w.getTextOfLine(currentJsLine+1, w.jsLineMap, w.jsFile.Content) // TODO: Strada is wrong here, we should be looking at `currentJsLine`, not `currentJsLine+1`
		if sw.prevEmittedCol < len(jsFileText)-1 {
			// There is remaining text on this line that will be part of next source span so write marker that continues
			sw.writeSourceMapMarkerEx(nil /*currentSpan*/, len(w.spansOnSingleLine), len(jsFileText)-1 /*endColumn*/, true /*endContinues*/)
		}

		// Emit Source text
		sw.iterateSpans(writeSourceMapSourceText)

		// Emit column number etc
		sw.iterateSpans(writeSpanDetails)

		w.sourceMapRecorder.WriteLine("---")
	}
}
